File: src/1-500/81/SearchRotatedSortedArray2.py
class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        if nums is None or len(nums) == 0:
            return False
        first, last = 0, len(nums) - 1

        while first <= last:
            mid = (first + last) // 2

            if nums[mid] == target:
                return True

            if nums[first] < nums[mid]:
                if nums[first] <= target and target < nums[mid]:
                    last = mid - 1

                else:
                    first = mid + 1

            elif nums[first] > nums[mid]:
                if nums[mid] < target and target <= nums[last]:
                    first = mid + 1

                else:
                    last = mid - 1

            else:
                first += 1

        return False

File: src/1-500/81/SearchRotatedSortedArray2.java
class Solution {

  public boolean search(int[] nums, int target) {
    if (nums == null || nums.length == 0) {
      return false;
    }

    int first = 0;
    int last = nums.length - 1;

    while (first <= last) {
      int mid = (first + last) / 2;

      if (nums[mid] == target) {
        return true;
      }

      if (nums[first] < nums[mid]) {
        if (nums[first] <= target && target < nums[mid]) {
          last = mid - 1;
        } else {
          first = mid + 1;
        }
      } else if (nums[first] > nums[mid]) {
        if (nums[mid] < target && target <= nums[last]) {
          first = mid + 1;
        } else {
          last = mid - 1;
        }
      } else {
        first++;
      }
    }

    return false;
  }
}
